L23-node-protobufjs-library.md

```diff
@@ -0,0 +1,72 @@
+ * @param {string} filename File path to the .proto file to load.
+ * @param {Object} options Options for loading the file and setting up the deserializers
+ * @return {PackageDefinition} An object containing definitions for all of the loaded services
+ */
+loadSync(filename, options)
+```
+
+The `options` parameter to each of those functions is the union of Protobuf.js's [`IParseOptions`](https://github.com/dcodeIO/protobuf.js/blob/cf7b26789f310dccf4c047c2e8ef5a3854f7f41e/index.d.ts#L1014) (which modifies how files are loaded) and [`IConversionOptions`](https://github.com/dcodeIO/protobuf.js/blob/cf7b26789f310dccf4c047c2e8ef5a3854f7f41e/index.d.ts#L1632) (which modifies deserialization functions), plus an `include` option to specify directories to search for dependencies. The full set of options is as follows:
+
+ - `keepCase`: a boolean indicating that field names should be preserved. The default is to change them to camel case.
+ - `longs`: Can be set to `Number` or `String` to indicate that long values should be represented as that type. The default is `Number`, or a safer object `Long` type if the library is installed.
+ - `enums`: Can be set to `String` have enums represented as strings. The default is to use their numeric value.
+ - `bytes`: Can be set to `Array` or `String` to indicate that bytes values should be represented as that type. The default is to use the Node built in type `Buffer`.
+ - `defaults`: a boolean indicating that default values should be set on output objects.
+ - `arrays`: a boolean indicating that empty arrays should be set for missing array values even if `defaults` is `false`.
+ - `objects`: a boolean indicating that empty objects should be set for missing object values even if `defaults` is `false`.
+ - `oneofs`: a boolean indicating that virtual oneof properties should be set to the present field's name.
+ - `json`: a boolean indicating that additional JSON compatibility conversions should be performed.
+ - `include`: an array of strings listing paths in which to search for imported `.proto` files.
+
+## Rationale
+
+This package provides very similar functionality to the existing `grpc.load` API, but with some differences. First, this API directly exposes all of the relevant Protobuf.js options instead of defining a subset of custom options. This package is explicitly a wrapper around Protobuf.js, so it makes more sense here to directly and consistently expose the relevant options from that library. Second, this API does not provide an equivalent to the `loadObject` function in the existing gRPC library. With the Protobuf.js options directly exposed, there is less benefit in allowing the user to separately load `.proto` files. Third, this adds an asynchronous API for loading `.proto` files. This has been requested before, and more closely matches existing Node APIs. It uses promises instead of callbacks because they are more versatile, and because Node is increasingly supporting them.
```

Comment:
Is it the plan that if protobuf.js introduces any breaking changes (e.g. renaming an option), this package will introduce a new release to make such an API change explicit to users?

Reply:
Yes, that is my intention.

Reply:
Maybe mention this in this section, that future breaking protobuf.js API changes will be contained in this pkg.